### PR TITLE
Add support for constrained routes from newer find-my-way

### PR DESCRIPTION
**Pull request by airhorns**
Closes https://github.com/fastify/fastify/issues/2498

### Checklist

- [X] run `npm run test` and `npm run benchmark`
- [X] tests and/or benchmarks are included
- [x] documentation is changed or added
- [X] commit message and code follows the [Developer's Certification of Origin](https://github.com/fastify/.github/blob/master/CONTRIBUTING.md#developers-certificate-of-origin-11)
      and the [Code of conduct](https://github.com/fastify/.github/blob/master/CODE_OF_CONDUCT.md)

This is the fastify side of https://github.com/delvedor/find-my-way/pull/170 which adds support for constraining routes to certain hostnames or using custom constraining strategies. 

It would need those bits of `find-my-way` to be merged and released before really being ready, but I'm opening this now for transparency and so anyone who wants to test can do so! Woop woop! 

No need to review quite yet, but if you want to use this feature, I'd love any help testing in your apps! To test, you'll have to import that branch of `find-my-way`, as well as this branch of `fastify`. `npm link` worked for me to do this.

### CODE REVIEW

File: lib/route.js
Comment (jsumners) [inline lib/route.js:85]: Unless we need an arrow function, I'd rather see `routing function () {}` or `routing () {}`.

Comment (airhorns) [reply lib/route.js:85]: Sorry yep this was just for debugging

**Review by mcollina — COMMENTED** [review]
good work! This is ready to go as the other PR is landed and released.

**Comment by Ethan-Arrowood** [comment]
@airhorns feel free to mark this PR as ready to review

**Comment by airhorns** [comment]
There's still a few more things I want to add like docs, and I was holding off until the `find-my-way` side got merged in case there were any major changes. I have been running this fork in production for a while now though so if you are wondering if it works, I think it does!

**Comment by fox1t** [comment]
Awesome work. I completely missed this PR in the first place and was about to open a feature request :)
Thanks @airhorns!

**Review by fox1t — APPROVED** [review]

**Review by Eomm — COMMENTED** [review]
We should update the server and the route docs accordingly

It would be useful to change the package.json to point to your branch airhorns:constraints in order to check the green CI too (it would be time spending checking all the node version and platforms 😅 )

**Comment by airhorns** [comment]
I think this is ready for review! It's hardcoded to use my fork of `find-my-way` to prove that the two sides agree, so I'm not going to undraft it until the find-my-way side is merged and I flip that back. Thanks for your feedback!

**Review by delvedor — CHANGES_REQUESTED** [review]
Great work! Probably [build/build-validation.js](https://github.com/fastify/fastify/blob/master/build/build-validation.js) should be updated as well.

**Review by jsumners — APPROVED** [review]

**Comment by airhorns** [comment]
> Great work! Probably build/build-validation.js should be updated as well.
 
 Ah, thanks for the heads up! I noticed there's no validation for the old `versioning` property in there right now -- is that on purpose? Is that just a bug where that wouldn't validate. If I am understanding correctly then, I'd need to add validation for both `versioning` and `constraints`? Also, the values of those config options may be instances of fancy objects as opposed to raw json scalars, is there a pattern for validating those kinds of things anywhere?

**Comment by delvedor** [comment]
`find-my-way` v4 [is out](https://github.com/delvedor/find-my-way/releases/tag/v4.0.0)! 🚀

**Comment by delvedor** [comment]
> I noticed there's no validation for the old versioning property in there right now -- is that on purpose?

I think it should be considered a bug.

We should validate that the user is passing the expected objects to `find-my-way`.

**Comment by airhorns** [comment]
Ok -- added validations for the old style and new style of passing constraints. `ajv` is a jsonschema validator, and these objects are real JS VM objects with functions and potentially prototypes and whatnot, so it's not really that strict a validation, but it will at least warn if there are missing keys or other weird stuff. Let me know if there is a better way to validate this kind of stuff!

**Review by mcollina — APPROVED** [review]
lgtm

**Comment by airhorns** [comment]
@delvedor is this config validation good enough or would you like to see something different?

**Comment by airhorns** [comment]
ping @delvedor

**Review by delvedor — APPROVED** [review]
LGTM

**Pull request merged**
